Add automation documentation tests for migration from sidecar to ambient

bmangoen opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

    1. If this is your first time, please read our contributor guidelines: https://github.com/istio-ecosystem/sail-operator/blob/main/CONTRIBUTING.md
    2. Discuss your changes before you start working on them. You can open a new issue in the [Sail Operator GitHub repository](https://github.com/istio-ecosystem/sail-operator/issues) or start a discussion in the [Sail Operator Discussion](https://github.com/istio-ecosystem/sail-operator/discussions). By this way, you can get feedback from the community and ensure that your changes are aligned with the project goals.
    3. If the PR is unfinished, make is as a draft.
-->

#### What type of PR is this?
<!--
In order to minimize the time taken to categorize your PR, add a label accoutring to the PR type defined above.
Please, use the following labels, according to the PR type:
    * Enhancement / New Feature - enhancement
    * Bug Fix - bug
    * Refactor - cleanup/refactor
    * Optimization - enhancement
    * Test - test-e2e
    * Documentation Update - documentation
-->

- [ ] Enhancement / New Feature
- [ ] Bug Fix
- [ ] Refactor
- [ ] Optimization
- [ ] Test
- [x] Documentation Update

#### What this PR does / why we need it:

Reorganized the migration from sidecar to ambient mode doc part and added automation tests for migration from sidecar to ambient mode

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
Add related issue or PR if exists.
-->
Fixes #1290

bmangoen committed:
Add automation test for migration from sidecar to ambient

Reorganized doc and added automation test for migration from sidecar to ambient mode

Signed-off-by: bmangoen <bmangoen@redhat.com>

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -194,20 +203,31 @@ For additional compatibility considerations refer to: link:../common/coexistence
 
 **Environment Validation:**
-[source,console]
+[source,bash,subs="attributes+"]
```
If you want to this command to be executed you will need to add this attribute also: `name="istio-migrate-from-sidecar-to-ambient"` with the same name of the test that we want to run

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -449,6 +500,10 @@ NAME      NAMESPACE   PROFILE   READY   STATUS    VERSION   AGE
 default   ztunnel               True    Healthy   v1.27.0   12m
```
```suggestion
default   ztunnel               True    Healthy   v{istio_latest_version}  12m
```

fjglira reviewed (commented):
Looks preatty good in general

bmangoen committed:
Add prebuilt functions to migration documentation

Adding prebuilt functions for creating default Istio resource, installing Bookinfo and creating Bookinfo Gateway

Signed-off-by: bmangoen <bmangoen@redhat.com>

istio-testing commented:
@bmangoen: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
e2e-kind-multicluster_sail-operator_main | 5d5f234e1942a57b7c945b0b881dea08b3e78a4e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/e2e-kind-multicluster_sail-operator_main/1994040183946743808) | true | `/test e2e-kind-multicluster`
e2e-kind-olm_sail-operator_main | 5d5f234e1942a57b7c945b0b881dea08b3e78a4e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/e2e-kind-olm_sail-operator_main/1994040184051601408) | true | `/test e2e-kind-olm`
e2e-kind_sail-operator_main | 5d5f234e1942a57b7c945b0b881dea08b3e78a4e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/e2e-kind_sail-operator_main/1994040184135487488) | true | `/test e2e-kind`
lint_sail-operator_main | b83b8bcf0688f604cae40d73c66a85da6522c5df | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/lint_sail-operator_main/1994041278710419456) | true | `/test lint`
docs-test_sail-operator_main | b83b8bcf0688f604cae40d73c66a85da6522c5df | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/docs-test_sail-operator_main/1994041278093856768) | false | `/test docs-test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bmangoen committed:
Remove steps which test external URL connectivity

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen committed:
Remove steps which test external URL connectivity

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen committed:
Variabilize istio latest version

* Variabilize istio latest version
* Remove execution of checking status

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen committed:
Fix all execution errors

* Remove unnecessary console characters / debug statements
* Fix all execution errors in the migration.adoc file

Signed-off-by: bmangoen <bmangoen@redhat.com>

fjglira commented:
Hey @bmangoen checking the execution after the latest changes I see this:

```
serviceaccount/bookinfo-productpage created
deployment.apps/productpage-v1 created
Warning: unrecognized format "int32"
Warning: unrecognized format "int64"
```
PLease check the applied yamls because seems that can be issues with the format

```
No resources found
label "istio.io/rev" not found.
namespace/bookinfo not labeled
```
Check also a step where is trying to label the namespace and get looking for a label

I think from that last error message is coming the issue that is causing to not go ambient mode correctly in the bookinfo namespace, because in the last step I see that the restarted bookinfo pods are still on sidecar mode:

```
deployment.apps/waypoint restarted
NAME                                      READY   STATUS              RESTARTS   AGE
bookinfo-gateway-istio-69b9bd5f75-b8rjg   1/1     Terminating         0          9s
bookinfo-gateway-istio-77d5b9c674-8mtfh   1/1     Running             0          3s
bookinfo-gateway-istio-7bcdc79466-5xrr8   0/1     ContainerCreating   0          0s
details-v1-5877f98768-625rj               2/2     Running             0          9s
details-v1-646bfdb84d-frdcm               0/2     Init:0/2            0          0s
productpage-v1-5b5b9f6c4d-c7zm8           2/2     Running             0          9s
productpage-v1-5c5797bc84-9mrq7           0/2     Init:0/2            0          0s
ratings-v1-6b9bdc976d-rhht5               0/2     Init:0/2            0          0s
ratings-v1-99cc4fbb4-vlq42                2/2     Running             0          9s
reviews-v1-75dd6854b4-ls7dw               0/2     Init:0/2            0          0s
reviews-v1-cc7784cd8-khflk                2/2     Running             0          9s
reviews-v2-78d775d888-lzq6q               2/2     Running             0          9s
reviews-v2-7f69f59fd8-vk6lp               0/2     Init:0/2            0          0s
reviews-v3-79c547f6c6-sl66g               0/2     Pending             0          0s
reviews-v3-7d45d95cb-rmb8w                2/2     Running             0          9s
waypoint-67c585948-mggcj                  1/1     Running             0          2s
Found pod(s) in namespace bookinfo: bookinfo-gateway-istio-69b9bd5f75-b8rjg
```

bmangoen committed:
Wait until istio-ca-root-cert propagation + fix KinD issue

Signed-off-by: bmangoen <bmangoen@redhat.com>

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -225,18 +243,28 @@ default   istio-system             1           1       1        default
 ----
 
-[source,console]
+[source,bash,subs="attributes+",name="istio-migrate-from-sidecar-to-ambient"]
 ----
 # Check current workloads with sidecar injection
 kubectl get pods -n bookinfo
+----
```
Checking this step in the output of the test I see that there is no sidecar injected in the bookinfo pods:

##### Test istio-migrate-from-sidecar-to-ambient: Step 6

NAME                                      READY   STATUS     RESTARTS   AGE
bookinfo-gateway-istio-75469d4894-th4xr   1/1     Running    0          6s
details-v1-6cc9f5cc44-qssj9               1/1     Running    0          9s
productpage-v1-7f885b46fc-s4srr           0/2     Init:0/2   0          5s
ratings-v1-77b8b6df5b-rvmvj               1/1     Running    0          9s
reviews-v1-fdbf79cd8-jx62n                1/1     Running    0          9s
reviews-v2-674c6d8b4-mtvtl                1/1     Running    0          9s
reviews-v3-7b775c7568-89lbh               0/2     Init:0/2   0          5s

fjglira replied:
Only one of the reviews has sidecar, so I think maybe there is a race condition between the install of the control plane and the installation of the bookinfo app

bmangoen replied:
Actually there is the [waiting pods for being ready in bookinfo namespace](https://github.com/bmangoen/sail-operator/blob/c8b5cd8f52ed19abd45e03a3c75d872d62499ef4/docs/migrate-from-sidecar-to-ambient/migration.adoc?plain=1#L278-L280).  
Maybe I should put it before `kubectl get pods -n bookinfo`, WDYT?

fjglira replied:
Yes, I think we should place a wait before the print just to be sure that we have everything set

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -476,15 +522,24 @@ kubectl get pods -n bookinfo
 Note: during the restart, sidecars will be updated to support HBONE while still functioning as traditional sidecars. Having HBONE support enabled is a prerequisite for ambient mode.
 
+ifdef::istio-migrate-from-sidecar-to-ambient[]
+with_retries wait_pods_ready_by_ns "bookinfo"
```
Add here also a print for the pods to check in the logs that all are 1/1 now again

bmangoen replied:
Done

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -476,15 +522,24 @@ kubectl get pods -n bookinfo
 ----
 # Check for HBONE is enabled in sidecar
-$ kubectl get pod $(kubectl get pods -n bookinfo -l app=ratings -o jsonpath='{.items[0].metadata.name}') -n bookinfo -o yaml | yq '.spec.containers[] | select(.name=="istio-proxy") | .env[] | select(.name=="PROXY_CONFIG")'
+kubectl get pod $(kubectl get pods -n bookinfo -l app=ratings -o jsonpath='{.items[0].metadata.name}') -n bookinfo -o yaml | yq '.spec.containers[] | select(.name=="istio-proxy") | .env[] | select(.name=="PROXY_CONFIG")'
```
This step is printing empty in the logs, so I think that there is returning empty:
##### Test istio-migrate-from-sidecar-to-ambient: Step 18


##### Test istio-migrate-from-sidecar-to-ambient: Step 19

fjglira replied:
So, there is no confirmation that HBONE is enable

bmangoen replied:
Right, I replaced using `yq` by a simple `grep` and we have the expected output

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -568,18 +634,13 @@ spec:
 
 Verify that waypoints are deployed correctly:
-[source,console]
+[source,bash,subs="attributes+",name="istio-migrate-from-sidecar-to-ambient"]
 ----
 kubectl get gateway waypoint -n bookinfo
 kubectl get deployment -n bookinfo -l gateway.networking.k8s.io/gateway-name=waypoint
```
There is no confirmation that gateway is up and running, in the logs we see:
```
##### Test istio-migrate-from-sidecar-to-ambient: Step 22

NAME       CLASS            ADDRESS   PROGRAMMED   AGE
waypoint   istio-waypoint             False        0s
NAME       READY   UP-TO-DATE   AVAILABLE   AGE
waypoint   0/1     1            0           0s
NAME                        READY   STATUS              RESTARTS   AGE
waypoint-7b7f57674d-97q6k   0/1     ContainerCreating   0          0s

##### Test istio-migrate-from-sidecar-to-ambient: Step 23
```

bmangoen replied:
Added `wait_pods_by_label "bookinfo" "gateway.networking.k8s.io/gateway-name=waypoint"` before displaying the `kubectl get` commands

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -1067,38 +1151,54 @@ kubectl rollout restart deployment -n bookinfo
-[source,console]
+[source,bash,subs="attributes+",name="istio-migrate-from-sidecar-to-ambient"]
 ----
 # Test authorization policies work through waypoint
```
We should validate that the output is the expected. because checking the logs I think that the output is different

```
##### Test istio-migrate-from-sidecar-to-ambient: Step 39

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 10.96.180.86:9080...
* Connected to reviews.bookinfo (10.96.180.86) port 9080 (#0)
> GET /reviews/1 HTTP/1.1
> Host: reviews.bookinfo:9080
> User-Agent: curl/7.88.1
> Accept: */*
> 
< HTTP/1.1 403 Forbidden
< content-length: 19
< content-type: text/plain
< date: Fri, 24 Apr 2026 09:38:53 GMT
< server: istio-envoy
< x-envoy-decorator-operation: reviews.bookinfo.svc.cluster.local:9080/*
< 
{ [19 bytes data]

100    19  100    19    0     0   4499      0 --:--:-- --:--:-- --:--:--  4750
* Connection #0 to host reviews.bookinfo left intact
RBAC: access denied
```

bmangoen replied:
I had to add the `cluster.local/ns/bookinfo/sa/bookinfo-ratings` in `rules.from.source.principals` to the `reviews-waypoint` AuthorizationPolicy for allowing the connectivity via ratings container

fjglira reviewed (commented):
I think we still need to add some wait and validations. At the end the migration looks almost good but there are some steps that seems to not be doing a proper validation

bmangoen committed:
Fix authorization policy for tests connectivity

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen commented:
> @bmangoen: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> docs-test_sail-operator_main	[29a171d](https://github.com/istio-ecosystem/sail-operator/commit/29a171d892a1c9945060682111febfba02951773)	[link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio-ecosystem_sail-operator/1379/docs-test_sail-operator_main/2048756222252814336)	false	`/test docs-test`
> Details

Needs the fix https://github.com/istio-ecosystem/sail-operator/pull/1890 to pass

fjglira commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -114,6 +114,13 @@ This migration guide uses example commands that can be adapted to any applicatio
 You can use any application that is currently running with sidecar injection enabled.
 
+ifdef::istio-migrate-from-sidecar-to-ambient[]
+create_default_istio
+install_bookinfo "{istio_release_name}"
+create_bookinfo_gateway_api "{istio_release_name}"
+with_retries wait_pods_ready_by_ns "bookinfo"
```
All this prerequisites are Ok, but the create default istio does not wait for the istiod pod creation and running and this causes bookinfo pod without sidecar injected:

```
##### Test istio-migrate-from-sidecar-to-ambient: Step 4

Found pod(s): istiod-6865c898f9-82r25
Waiting for pod istiod-6865c898f9-82r25 in namespace istio-system to be ready...
pod/istiod-6865c898f9-82r25 condition met
Found 1 istiod pods in istio-system namespace
NAME      NAMESPACE      PROFILE   REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION           AGE
default   istio-system             1           1       1        default           Healthy   v1.30.0-alpha.2   7s
NAME                      READY   STATUS    RESTARTS   AGE
istiod-6865c898f9-82r25   1/1     Running   0          7s
NAME      NAMESPACE      PROFILE   READY   STATUS    IN USE   VERSION           AGE
default   istio-system             True    Healthy   True     v1.30.0-alpha.2   7s
No resources found

##### Test istio-migrate-from-sidecar-to-ambient: Step 5

NAME       STATUS   AGE
bookinfo   Active   7s

##### Test istio-migrate-from-sidecar-to-ambient: Step 6

NAME                                      READY   STATUS            RESTARTS   AGE
bookinfo-gateway-istio-75469d4894-5488q   1/1     Running           0          4s
details-v1-764c46cfdb-bgnvq               1/1     Running           0          7s
productpage-v1-85664dccbc-x8tj2           1/2     PodInitializing   0          3s
ratings-v1-779fdc7f86-r9z7h               1/1     Running           0          7s
reviews-v1-85964f9f98-4tzf4               1/1     Running           0          7s
reviews-v2-6f7fbdc6fd-gstgh               1/1     Running           0          7s
reviews-v3-689b477554-s64jm               1/1     Running           0          7s
```

We should have a validation step when we print the bookinfo pods to validate that they are 2/2 or have sidecar injected to fail the test

fjglira replied:
I think inside the `create_default_istio` we should place a wait to get the istiod ready before installing the bookinfo app, and add a validation to check that bookinfo pods are 2/2

bmangoen replied:
Added

fjglira commented on common/scripts/kind_provisioner.sh:
```diff
@@ -203,6 +203,12 @@ function setup_kind_cluster() {
+  echo "Increasing inotify limits on KIND node..."
+  docker exec "${NAME}"-control-plane sysctl -w fs.inotify.max_user_instances=8192 || true
+  docker exec "${NAME}"-control-plane sysctl -w fs.inotify.max_user_watches=524288 || true
+
```
These changes are going to be overriden because we get these files from istio/common-files repo

bmangoen replied:
Ah you're right ! Do you think guys from the community would accept this change upstream?

fjglira replied:
I think we should give a try to do the change upstream. I don't see any downside on this

bmangoen replied:
Removed and waiting for https://github.com/istio/common-files/pull/1278 to be accepted and merged

fjglira replied:
So, this errors:

```
##### Test istio-migrate-from-sidecar-to-ambient: Step 28

Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name


Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name



##### Test istio-migrate-from-sidecar-to-ambient completed: FAIL
```

Are they related to that change?

fjglira replied:
Seems that Jonh is concerned about being masking an actual leak instead of a resource consumption issue

bmangoen replied:
> So, this errors:
> 
> ```
> ##### Test istio-migrate-from-sidecar-to-ambient: Step 28
> 
> Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
> 	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
> 	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name
> 
> 
> Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
> 	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
> 	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name
> 
> 
> 
> ##### Test istio-migrate-from-sidecar-to-ambient completed: FAIL
> ```
> 
> Are they related to that change?

I don't think so. I'm looking at it

bmangoen replied:
> Seems that Jonh is concerned about being masking an actual leak instead of a resource consumption issue

Yeah he's right ! So we need to find the leak if there is (or maybe it's just in my local environment, I need to confirm this)

fjglira reviewed (commented):
Just some small comment the rest of the test looks ok, the connectivity is working and the migration seems successfully

bmangoen committed:
Add wait step at Istio creation

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen commented:
/retest

fjglira commented:
Ok, checking these errors:

```
##### Test istio-migrate-from-sidecar-to-ambient: Step 28

Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
	* invalid policy productpage-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name


Error from server: error when creating "STDIN": admission webhook "validation.istio.io" denied the request: configuration is invalid: 2 errors occurred:
	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `value` for `key` source.workload_name: unknown attribute: source.workload_name
	* invalid policy reviews-ztunnel-protection.bookinfo: invalid `notValue` for `key` source.workload_name: unknown attribute: source.workload_name
```

These error happens on this step:
```
[source,bash,subs="attributes+",name="istio-migrate-from-sidecar-to-ambient"]
----
cat <<EOF | kubectl apply -f-
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
metadata:
  name: productpage-ztunnel-protection
  namespace: bookinfo
spec:
  targetRefs:
  - kind: Service
    group: ""
    name: productpage
  action: DENY
  rules:
  - when:
    - key: source.workload_name
      notValues: ["waypoint"]
```

But I don't think are related to the step it self, seems to be related to a kind cluster issue, checking previous steps I see some pods on error state. Maybe worth include some debugging steps to check what is going on. But don't see to be related to the test it self

bmangoen committed:
Fix correct label

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen committed:
Fix ztunnel authorization policies

Signed-off-by: bmangoen <bmangoen@redhat.com>

bmangoen commented on docs/migrate-from-sidecar-to-ambient/migration.adoc:
```diff
@@ -870,11 +952,11 @@ spec:
     group: ""
     name: productpage
-  action: DENY
+  action: ALLOW
```
@fjglira I modified a bit the AuthorizationPolicy. They were too restrictive and it blocks all legitimate service-to-service traffic (like ratings → reviews)

bmangoen commented:
/retest

bmangoen commented:
/test integ-tests-arm64

dgn commented:
/test all
